class.c

```diff
@@ -636,184 +636,264 @@ Perl_class_apply_attributes(pTHX_ HV *stash, OP *attrlist)
     op_free(attrlist);
 }
 
+/*
+
+Called when a compilation failure occurs when defining a class.
+
+Returns the given stash to a clean state, as if none of the class has
+been defined so a new attempt can be made.
+
+*/
+
+static void
+S_class_cleanup_definition(pTHX_ HV *stash) {
+    struct xpvhv_aux *aux = HvAUX(stash);
+
+    SvREFCNT_dec(aux->xhv_class_superclass);
+    aux->xhv_class_superclass = NULL;
```

Comment:
I suggest lifting the pointers to a C autos, and setting struct field to NULL, then call ```SvREFCNT_dec```. Analyse if ``` SvREFCNT_dec_NN``` is appropriate.